source/PlanetLabel.cpp

```diff
@@ -91,76 +82,135 @@ PlanetLabel::PlanetLabel(const Point &position, const StellarObject &object, con
 	else if(planet.GetGovernment())
 	{
 		government = "(" + planet.GetGovernment()->GetName() + ")";
-		color = planet.GetGovernment()->GetColor();
-		color = Color(color.Get()[0] * .5f + .3f, color.Get()[1] * .5f + .3f, color.Get()[2] * .5f + .3f);
+		color = Color::Combine(.5f, planet.GetGovernment()->GetColor(), 1.f, Color(.3f));
 		if(!planet.CanLand())
 			hostility = 3 + 2 * planet.GetGovernment()->IsEnemy();
 	}
 	else
 	{
-		color = Color(.3f, .3f, .3f, 1.f);
 		government = "(No government)";
+		color = Color(.3f);
 	}
-	float alpha = static_cast<float>(min(.5, max(0., .6 - (position.Length() - object.Radius()) * .001 * zoom)));
-	color = Color(color.Get()[0] * alpha, color.Get()[1] * alpha, color.Get()[2] * alpha, 0.);
-
-	if(!system)
-		return;
-
-	// Figure out how big the label has to be.
-	double width = max(FontSet::Get(18).Width(name), FontSet::Get(14).Width(government)) + 8.;
 
-	// Try to find a label direction that not overlapping under any zoom.
-	for(int d = 0; d < 4; ++d)
-		if(!Overlaps(*system, object, Preferences::MinViewZoom(), width, d)
-				&& !Overlaps(*system, object, Preferences::MaxViewZoom(), width, d))
+	// Figure out how big the label is.
+	const Font &font = FontSet::Get(14);
+	const Font &bigFont = FontSet::Get(18);
+	const double labelWidth = max(bigFont.Width(name), font.Width(government));
+	const double nameHeight = bigFont.Height();
+	const double labelHeight = nameHeight + (government.empty() ? 0. : 1. + font.Height());
+	const Point labelDimensions = {labelWidth + BORDER * 2., labelHeight + BORDER * 2.};
+
+	// Try to find a label direction that is not overlapping under any zoom.
+	const double minZoom = Preferences::MinViewZoom();
+	const double maxZoom = Preferences::MaxViewZoom();
+	for(const double angle : LINE_ANGLES)
+	{
+		SetBoundingBox(labelDimensions, angle);
+		if(!HasOverlaps(labels, system, object, minZoom) && !HasOverlaps(labels, system, object, maxZoom))
 		{
-			direction = d;
-			return;
+			innerAngle = angle;
+			break;
 		}
+	}
 
-	// If we can't find a suitable direction, then try to find a direction under the current
-	// zoom that is not overlapping.
-	for(int d = 0; d < 4; ++d)
-		if(!Overlaps(*system, object, zoom, width, d))
-		{
-			direction = d;
-			return;
-		}
+	// No non-overlapping choices, so set this to the default.
```

Comment:
"No non-overlapping choices"  only at every zoom level. Why did you choose to set it at the first angle, instead of searching for a non-overlapping choice at the current zoom level (like what the original code did)?

Reply:
A few reasons.  One, we're now doing this at system entry time, and the zoom level might not be what you'd expect (like during the zoom-out when departing a planet).  Two, the increased number of searched angles makes it rather unlikely for the case to come up.  And three, it was never a really good backup plan, because even if it did find something (which was unlikely to begin with), there'd be an overlap as soon as the zoom level changed anyway.

A better approach would be to rate each of the overlaps by amount of overlap, and choose the one with the least overlap.  Still not great, but if you're going to throw cpu time at it, better than going through all the objects and labels again.